📖 Update Documentation (#170)
kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm: README.md

* [cluster-api.sigs.k8s.io](https://cluster-api.sigs.k8s.io)
* [The Kubebuilder Book](https://book.kubebuilder.io)

> ## How to build, deploy and test CAPBK

Reviewer: this might be the worst acronym we've ever created

> CAPBK is built using kubebuilder, please refer to the kubebuilder docs for build and deploy in a cluster.

Reviewer: link to the kubebuilder docs you are recommending they read

There are many ways to achieve this, but a convenient way is to use [kind](https://kind.sigs.k8s.io/) and tools like
[tilt](https://tilt.dev/) or [skaffold](ttps://skaffold.dev/); a minimal example of Tiltfile is.

```
allow_k8s_contexts('kubernetes-admin@kubernetes')

controllers = {
    'capi': {
        'path': '../cluster-api',
        'image': 'gcr.io/k8s-staging-cluster-api/cluster-api-controller:dev',
    },
    'cabpk': {
        'path': './',
        'image': 'gcr.io/k8s-staging-cluster-api/cluster-api-bootstrap-provider-kubeadm:dev',
    },
    'capd': {
        'path': '../cluster-api-provider-docker',
        'image': 'gcr.io/k8s-staging-cluster-api/cluster-api-provider-docker:dev',
    },
}

for name, controller in controllers.items():
    command = '''sed -i '' -e 's@image: .*@image: '"{}"'@' ./{}/config/default/manager_image_patch.yaml'''.format(controller['image'], controller['path'])
    local(command)

    k8s_yaml(local('kustomize build ' + controller['path'] + '/config/default'))

    docker_build(controller['image'], controller['path'])

```

> See https://github.com/chuckha/capi-dev for an example of a more complex developemt environment using [tilt](https://tilt.dev/).

Reviewer: do you want to add a real link to `capi-dev` here?

## How it works?
Once your test environment is in place, create a `Cluster` object and its corresponding `DockerCluster`
infrastructure object.

```yaml
kind: DockerCluster
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-cluster-docker
---
kind: Cluster
apiVersion: cluster.x-k8s.io/v1alpha2
metadata:
  name: my-cluster
spec:
  infrastructureRef:
    kind: DockerCluster
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    name: my-cluster-docker
```

> After you can start creating machines by defining a `Cluster`, its corresponding `DockerMachine` object, and

Reviewer: ```suggestion
Now you can start creating machines by defining a `Machine`, its corresponding `DockerMachine` object, and
```

the `KubeadmConfig` bootstrap object.

```yaml
kind: KubeadmConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-control-plane1-config
spec:
  initConfiguration:
    nodeRegistration:
      kubeletExtraArgs:
        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
  clusterConfiguration:
    controllerManager:
      extraArgs:
        enable-hostpath-provisioner: "true"
    networking:
      podSubnet: "10.244.0.0/16"
      dnsDomain: ""
      serviceSubnet: ""
---
kind: DockerMachine
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-control-plane1-docker
---
kind: Machine
apiVersion: cluster.x-k8s.io/v1alpha2
metadata:
  name: my-control-plane1
  labels:
    cluster.x-k8s.io/cluster-name: my-cluster
    cluster.x-k8s.io/control-plane: "true"
    set: controlplane
spec:
  bootstrap:
    configRef:
      kind: KubeadmConfig
      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
      name: my-control-plane1-config
  infrastructureRef:
    kind: DockerMachine
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    name: my-control-plane1-docker
  version: "v1.14.2"
```

> CAPBK main responsibility is to convert a `KubeadmConfig` bootstrap object into a cloud-init script that is

Reviewer: ```suggestion
CAPBK's main responsibility is to convert a `KubeadmConfig` bootstrap object into a cloud-init script that is
```

> going to turn a Machine into a Kubernetes Node using kubeadm.

Reviewer: ```suggestion
going to turn a Machine into a Kubernetes Node using `kubeadm`.
```

> The cloud-init script will be saved into the `KubeadmConfig.Status.BootstrapData` and then the infrastructure provider

Reviewer: is it a cloud-init script or a cloud-config? I don't really know the right vocabulary here. @detiber ?

(CAPD in this example) will pick up this value and proceed with the machine creation and the actual bootstrap.

#### KubeadmConfig examples

Bootstrap control plane node:
```yaml
kind: KubeadmConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-control-plane1-config
spec:
  initConfiguration:
    nodeRegistration:
      kubeletExtraArgs:
        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
  clusterConfiguration:
    controllerManager:
      extraArgs:
        enable-hostpath-provisioner: "true"
    networking:
      podSubnet: "10.244.0.0/16"

>       dnsDomain: ""

Reviewer: Outside of the scope of this PR -- Should we make these optional or does an empty string indicate something different than nil?

Author: this is probably a leftover of https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/156
cc @vincepri

      serviceSubnet: ""
```

Additional control plane nodes:
```yaml
kind: KubeadmConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-control-plane2-config
spec:
  joinConfiguration:
    nodeRegistration:
      kubeletExtraArgs:
        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
    controlPlane: {}
```

worker nodes:
```yaml
kind: KubeadmConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
metadata:
  name: my-worker1-config
spec:
  joinConfiguration:
    nodeRegistration:
      kubeletExtraArgs:
        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
```

### Simplifying KubeadmConfig objects
`KubeadmConfig` object allows full control of Kubeadm init/join operations by exposing raw `InitConfiguration`,
`ClusterConfiguration` and `JoinConfiguration` objects.

> valid combinations of configuration objects are:
> - at least one of `InitConfiguration` and `ClusterConfiguration` for the first control plane node only

> > - `JoinConfiguration` for worker nodes or additional control plane nodes

Reviewer: ```suggestion
> - `JoinConfiguration` for worker nodes and additional control plane nodes
```

>
> see [examples](#kubeadmconfig-examples)

By design CABPK respect all the user settings in configuration `KubeadmConfig`; in case some of those settings are missing,
CABPK applies the following set of rules that can simplify the preparation of `KubeadmConfig`:

> - if `KubeadmConfig.clusterConfiguration.clusterName` is empty, CABPK will automatically set this value to

Reviewer: What do you think about having a table for this kind of information? Something like:

CABPK will fill in some values if they are left empty with sensible defaults:

Field | Default
---|---
`KubeadmConfig.clusterConfiguration.clusterName` | `Cluster.Name`
... | ...

`Cluster.metadata.name` value.
- if `KubeadmConfig.clusterConfiguration.controlPlaneEndpoint` is empty, CABPK will automatically set this value to
the value in the first entry in `Cluster.status.apiEndpoints`; this values is set by the infrastructure provider
when creating the cluster load balancer or, in some cases, when creating the first control plane machine.
- if `KubeadmConfig.clusterConfiguration.networking.serviceSubnet` is empty, CABPK will automatically set this value to
`Cluster.spec.clusterNetwork.serive.cidrBlocks` value. Same apply for the other networking settings.
- if `KubeadmConfig.joinConfiguration.discovery` is not set, CABPK will automatically create a short lived BootstrapToken
in your cluster, and then inject corresponding settings into the `JoinConfiguration` object.

### Bootstrap Orchestration
CABPK orchestrates the generation of cloud-init scripts of different machines in order to ensure a cluster
bootstrap process that will be compliant with the correct Kubeadm init/join sequence. More in detail:
1. cloud-init generation starts only after `Cluster.InfrastructureReady` flag is set to `true`.

> 2. at this stage, the cloud-init for **one control plane machine only** is generated (kubeadm init).

Reviewer: It might be nice to say we support multiple control plane machines initing at the same time.

Perhaps don't bold the one control plane machine part and say something like

`the cloud-init script/cloud-config will be generated for the first control plane machine even if multiple control plane machines are ready`

> 3. after `Cluster.Annotations[cluster.x-k8s.io/control-plane-ready]` is set to true, the cloud-init

Reviewer: This should be `Cluster.metadata.annotations` or the paragraph above should be switched to `Cluster.Name`, but we should be consistent with how we reference the metadata fields
